### PR TITLE
[Merged by Bors] - [DEV FEATURE] Deterministic long lived subnets

### PR DESCRIPTION
**Pull request by divagant-martian**
## Issue Addressed

#2847 

## Proposed Changes
Add under a feature flag the required changes to subscribe to long lived subnets in a deterministic way

## Additional Info

There is an additional required change that is actually searching for peers using the prefix, but I find that it's best to make this change in the future

### CODE REVIEW

**Comment by AgeManning** [comment]
Is this ready for review?

File: beacon_node/network/src/subnet_service/attestation_subnets.rs [hunk 1]
```diff
@@ -145,31 +161,83 @@ impl<T: BeaconChainTypes> AttestationService<T> {
...
 
+    #[cfg(feature = "deterministic_long_lived_attnets")]
+    pub fn new(
+        beacon_chain: Arc<BeaconChain<T>>,
+        node_id: ethereum_types::U256,
+        config: &NetworkConfig,
+        log: &slog::Logger,
+    ) -> Self {
+        let log = log.new(o!("service" => "attestation_service"));
+
```
Comment (AgeManning) [inline beacon_node/network/src/subnet_service/attestation_subnets.rs:172]: Maybe we should an an INFO log saying something like:
"Deterministic subnets enabled; "subnets_per_node" => spec.subnets_per_node

**Review by AgeManning — APPROVED** [review]
This looks really good to me. Keen to try it out :)

File: beacon_node/network/src/subnet_service/attestation_subnets.rs (hunk 1 above)
Comment (divagant-martian) [reply beacon_node/network/src/subnet_service/attestation_subnets.rs:172]: done in https://github.com/sigp/lighthouse/pull/3453/commits/97aa24aa4b3a41d1882cd407cb05dccce99774bc

**Comment by michaelsproul** [comment]
Once the merge conflict is fixed and the dust has settled from v3.1.2, we can merge :rocket:

**Comment by divagant-martian** [comment]
don't think we want this merge tho, @AgeManning ?

**Comment by AgeManning** [comment]
Yeah I think we do want this, as long as the default is the usual behaviour. 

The compile-time flag enables this extra functionality and is handy for further testing

**Comment by divagant-martian** [comment]
@AgeManning this is what's been running in wood (so, without the feature on) all this time and to me it looks good

**Pull request closed without merging**